### PR TITLE
Issue 305 bulk insert

### LINKED ISSUE #304: Provide get by many ids routes
We should provide some route to which, providing a list of ids returns a list of entities.

See discussion in #302

### LINKED ISSUE #305: Routes to post multiple entities instead of one
For each entity type, we should have a route to add a collection of that kind of entity.  This would reduce HTTP roundtrip overhead and speed up imports.

### PR DESCRIPTION
**Pull request by yogsototh**
I made one commit per route, tell me if you prefer only one big squashed commit.

Also, I'm not fond of the route naming I choose. For now, it is: `/:entity/multi`

I didn't want to call the route `bulk` because the current implementation focus on reducing the number of HTTP request not really on being able to handle big volume of insertions, yet.

Closes #305
Closes #304

### CODE REVIEW

**Comment by saintx** [comment]
One big squashed commit is best in nearly every case.  The reason is to facilitate potential rollback.  It's highly unlikely we will ever need to roll back part of this, but not the rest of it.

WRT naming, in general, I like [/actor & /actors], [/campaign & /campaigns] etc for single & multi, and not the sub-route.  The name of the route should not be chosen based on implementation details (such as whether the implementation is optimized for large batches or small ones), because the user will not and should not need to know or concern themselves with those details.  The *only* thing the user needs to know, is that you use "/actor" for a single entity, and "/actors" for multiple entities, in my opinion.  This keeps it simple, and lets us choose how we'll optimize under the hood later on.

**Comment by yogsototh** [comment]
as we should also provide routes returning a list of entities provided a list of ids, I'm following your proposal to use "/actors"

And I'll squash all my commit together.

**Comment by gbuisson** [comment]
LGTM (Looks Good To Me)

**Comment by polygloton** [comment]
maybe we could have a different top level route, something like /ctia/entities (POST & GET)

**Comment by polygloton** [comment]
Also I think there wold only need to be one route.  We can allow any mix of entity types to be posted if require that the :type filed be set.

**Comment by polygloton** [comment]
We should have a configured maximum number of entities that may be posted.  Its the kind of thing you want to fail when someone is doing their integration testing so that they adjust their client implementation.  I think it should be a property.  This is the kind of thing that could cause production issues down the road if there is no maximum.

**Comment by saintx** [comment]
Agreed.  I found myself contemplating "Why stop at 1k?  Why not 10k?  Why not 100k?  What happens?"

**Comment by gbuisson** [comment]
You might also want to enforce max content size in order to prevent upload bombs.

**Comment by saintx** [comment]
I agree that guarding against upload bombs is good, but at some point, I want us to actually merge #305 and stop tacking new features onto it.  I need this thing.  New features and improvements can go into their own issue.

File: src/ctia/http/routes/bulk.clj [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    :actor          ent/realize-actor
+    :campaign       ent/realize-campaign
+    :coa            ent/realize-coa
+    :exploit-target ent/realize-exploit-target
+    :feedback       ent/realize-feedback
+    :incident       ent/realize-incident
+    :indicator      ent/realize-indicator
+    :judgement      ent/realize-judgement
+    :sighting       ent/realize-sighting
+    :ttp            ent/realize-ttp))
```
Comment (polygloton) [inline src/ctia/http/routes/bulk.clj]: This can go into ctia.domain.entities now.

File: src/ctia/http/routes/bulk.clj [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+                      :read-campaign
+                      :read-coa
+                      :read-exploit-target
+                      :read-feedback
+                      :read-incident
+                      :read-indicator
+                      :read-judgement
+                      :read-sighting
+                      :read-ttp}
+      (let [bulk (into {} (filter (comp not empty? second)
```
Comment (polygloton) [inline src/ctia/http/routes/bulk.clj]: you could use "seq" instead of "not empty?" I think

File: src/ctia/http/routes/bulk.clj [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+            [ring.util.http-response :refer :all]
+            [schema.core :as s]
+            [clojure.tools.logging :as log]))
+
+(defn singular [k]
+  "remove the last s of a keyword see test for an example."
+  (-> k
+      name
+      (str/replace #"s$" "")
+      keyword))
```
Comment (polygloton) [inline src/ctia/http/routes/bulk.clj]: Maybe move this to a lib?  Not sure what it would be.  Perhaps ctia.lib.keyword.  Also move tests.  :)

**Comment by polygloton** [comment]
Looks good.  I'm merging this.  Minor nits may be addressed in another PR.  :)

File: src/ctia/http/routes/bulk.clj (hunk 2 above)
Comment (AlexBaranosky) [inline src/ctia/http/routes/bulk.clj]: I prefer `(remove (comp empty? val))` :)

File: src/ctia/http/routes/bulk.clj (hunk 1 above)
Comment (AlexBaranosky) [inline src/ctia/http/routes/bulk.clj]: @yogsototh `case` is both more efficient, and less code.

File: src/ctia/http/routes/bulk.clj (hunk 2 above)
Comment (yogsototh) [inline src/ctia/http/routes/bulk.clj]: double Linting 👍 :)

**Pull request merged**